### ISSUE TITLE
Telit Terminal IO profile?

### ISSUE THREAD

**Issue by jcrabanal** [kai-morich/SimpleBluetoothLeTerminal#2]
Hello kai, thanks for directing me here from the Play Store app page.

This particular BLE device I'm working on uses a Telit chipset. This simplified version of your Play Store app does not seem to include the profile for these, however your app on the Play Store works flawlessly.

I've tried adding Telit's UUIDs:
```
private static final UUID BLUETOOTH_LE_TELIT_SERVICE = UUID.fromString("0000fefb-0000-1000-8000-00805f9b34fb");
    private static final UUID TELIT_DATA_RX = UUID.fromString("00000001-0000-1000-8000-008025000000");
    private static final UUID TELIT_DATA_TX = UUID.fromString("00000002-0000-1000-8000-008025000000");
```

```
 @Override
    public void onServicesDiscovered(BluetoothGatt gatt, int status) {
        Log.d(TAG, "servicesDiscovered, status " + status);
        if (canceled) {
            return;
        }
        writePending = false;
        for (BluetoothGattService gattService : gatt.getServices()) {
            UUID uuid = gattService.getUuid();
            if (uuid.equals(BLUETOOTH_LE_CC254X_SERVICE)) {
                Log.d(TAG, "service cc254x uart");
                //for(BluetoothGattCharacteristic characteristic : gattService.getCharacteristics())
                //    Log.d(TAG, "characteristic "+characteristic.getUuid());
                readCharacteristic = gattService.getCharacteristic(BLUETOOTH_LE_CC254X_CHAR_RW);
                writeCharacteristic = gattService.getCharacteristic(BLUETOOTH_LE_CC254X_CHAR_RW);
            } else if (uuid.equals(BLUETOOTH_LE_NRF_SERVICE)) {
                Log.d(TAG, "service nrf uart");
                //for(BluetoothGattCharacteristic characteristic : gattService.getCharacteristics())
                //    Log.d(TAG, "characteristic "+characteristic.getUuid());
                BluetoothGattCharacteristic rw2 = gattService.getCharacteristic(BLUETOOTH_LE_NRF_CHAR_RW2);
                BluetoothGattCharacteristic rw3 = gattService.getCharacteristic(BLUETOOTH_LE_NRF_CHAR_RW3);
                if (rw2 != null && rw3 != null) {
                    int rw2prop = rw2.getProperties();
                    int rw3prop = rw3.getProperties();
                    boolean rw2write = (rw2prop & BluetoothGattCharacteristic.PROPERTY_WRITE) != 0;
                    boolean rw3write = (rw3prop & BluetoothGattCharacteristic.PROPERTY_WRITE) != 0;
                    Log.d(TAG, "characteristic properties " + rw2prop + "/" + rw3prop);
                    if (rw2write && rw3write) {
                        onSerialConnectError(new IOException("multiple write characteristics (" + rw2prop + "/" + rw3prop + ")"));
                        return;
                    } else if (rw2write) { // some devices use this ...
                        writeCharacteristic = rw2;
                        readCharacteristic = rw3;
                    } else if (rw3write) { // ... and other devices use this characteristic
                        writeCharacteristic = rw3;
                        readCharacteristic = rw2;
                    } else {
                        onSerialConnectError(new IOException("no write characteristic (" + rw2prop + "/" + rw3prop));
                        return;
                    }
                }
            } else if (uuid.equals(BLUETOOTH_LE_TELIT_SERVICE)) {
                Log.d(TAG, "service cc254x uart");
                //for(BluetoothGattCharacteristic characteristic : gattService.getCharacteristics())
                //    Log.d(TAG, "characteristic "+characteristic.getUuid());
                readCharacteristic = gattService.getCharacteristic(TELIT_DATA_RX);
                writeCharacteristic = gattService.getCharacteristic(TELIT_DATA_TX);
            }
        }
        if (readCharacteristic == null || writeCharacteristic == null) {
            for (BluetoothGattService gattService : gatt.getServices()) {
                Log.d(TAG, "service " + gattService.getUuid());
            }
            onSerialConnectError(new IOException("no serial profile found"));
            return;
        }
        int writeProperties = writeCharacteristic.getProperties();
        if ((writeProperties & (BluetoothGattCharacteristic.PROPERTY_WRITE +     // Microbit,HM10-clone have WRITE
                BluetoothGattCharacteristic.PROPERTY_WRITE_NO_RESPONSE)) == 0) { // HM10,TI uart have only WRITE_NO_RESPONSE
            onSerialConnectError(new IOException("write characteristic not writable"));
            return;
        }
        if (!gatt.setCharacteristicNotification(readCharacteristic, true)) {
            onSerialConnectError(new IOException("no notification for read characteristic"));
            return;
        }
        BluetoothGattDescriptor readDescriptor = readCharacteristic.getDescriptor(BLUETOOTH_LE_CCCD);
        if (readDescriptor == null) {
            onSerialConnectError(new IOException("no CCCD descriptor for read characteristic"));
            return;
        }
        int readProperties = readCharacteristic.getProperties();
        if ((readProperties & BluetoothGattCharacteristic.PROPERTY_INDICATE) != 0) {
            Log.d(TAG, "enable read indication");
            readDescriptor.setValue(BluetoothGattDescriptor.ENABLE_INDICATION_VALUE);
        } else if ((readProperties & BluetoothGattCharacteristic.PROPERTY_NOTIFY) != 0) {
            Log.d(TAG, "enable read notification");
            readDescriptor.setValue(BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE);
        } else {
            onSerialConnectError(new IOException("no indication/notification for read characteristic (" + readProperties + ")"));
            return;
        }
        Log.d(TAG, "writing read characterictic descriptor");
        if (!gatt.writeDescriptor(readDescriptor)) {
            onSerialConnectError(new IOException("read characteristic CCCD descriptor not writable"));
        }
        Log.d(TAG, "writing read characteristic descriptor");
        // continues asynchronously in onDescriptorWrite()
    }
```

Am I doing anything wrong?

EDIT: Nevermind. I figured out what the "credits" characteristic is for.

For anyone reading this, the credits are needed for sending/receiving data. If you don't send enough credits on the characteristic with UUID:

``
private static final UUID RX_CREDITS = UUID.fromString("00000003-0000-1000-8000-008025000000");
``
With this code:
```
private void trySend(BluetoothGattCharacteristic characteristic, int nCredits) {
        boolean bResult = characteristic.setValue(nCredits, BluetoothGattCharacteristic.FORMAT_UINT8, 0);
        Log.v("ble", "trySendCredits setValue(): " + bResult);
        bResult = btGatt.writeCharacteristic(characteristic);
        Log.v("ble", "trySendCredits writeCharacteristic(): " + bResult);
    }
```

Then nothing will work.

**Comment by kai-morich** [comment]
with Telit devices the communication is more complex. The credit characteristics have to be initialized in the right order and after send/receive you have renew credits. To not bload the already complex bluetooth communication, I did not include the Telit protocol

**Comment by The-Mekon** [comment]
I've also been struggling to get Telit's Terminal I/O to work.  The only documentation I've found seems rather vague.  Is there any detailed source for this, or is it a case of trial and error?

**Comment by jcrabanal** [comment]
Yeah, the documentation is pretty scarce and android BLE APIs are pretty shitty to begin with. Here you have some sample code. Start with the connectSync() or connectAsync() methods.

```
import android.annotation.SuppressLint;
import android.annotation.TargetApi;
import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothGatt;
import android.bluetooth.BluetoothGattCallback;
import android.bluetooth.BluetoothGattCharacteristic;
import android.bluetooth.BluetoothGattDescriptor;
import android.bluetooth.BluetoothGattService;
import android.bluetooth.BluetoothProfile;
import android.content.Context;
import android.os.Build;
import android.os.Looper;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.TextUtils;
import android.util.Log;

import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.util.Arrays;
import java.util.List;
import java.util.UUID;
import java.util.concurrent.TimeUnit;

@SuppressWarnings({"CharsetObjectCanBeUsed", "unused"})
@TargetApi(Build.VERSION_CODES.JELLY_BEAN_MR2)
@SuppressLint("ObsoleteSdkInt")
public class BleSerialPortTelitGattCallback extends BluetoothGattCallback {

    private static final String TAG = "BleSerialPort";
    private String sMacAddress;
    private long nConnectTimeout;
    private boolean bAutoConnect;

    private enum TelitUuid {
        Service,
        Rx,
        Tx,
        RxCredits,
        TxCredits,
        TxAndTxCredits
    }

    @NonNull
    private final BluetoothAdapter btAdapter;
    @NonNull
    private final Context context;
    private BluetoothGatt btGatt;
    private BluetoothGattCharacteristic charRx;
    private BluetoothGattCharacteristic charTx;
    private BluetoothGattCharacteristic charRxCredits;
    private BluetoothGattCharacteristic charTxCredits;
    private boolean bConnected;
    private boolean bConnectOperationFinished;
    private int nReceivedCredits;

    public BleSerialPortTelitGattCallback(@NonNull Context context, @NonNull BluetoothAdapter btAdapter) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.JELLY_BEAN_MR2) {
            throw new UnsupportedOperationException("BleSerialPortTelitGattCallback is not available on API levels below 18 (Android 4.3)");
        }
        this.btAdapter = btAdapter;
        this.context = context.getApplicationContext();
    }

    private UUID toUuid(TelitUuid telitUuid) {
        switch (telitUuid) {
            case Service:
                return UUID.fromString("0000fefb-0000-1000-8000-00805f9b34fb");
            case Rx:
                return UUID.fromString("00000001-0000-1000-8000-008025000000");
            case Tx:
                return UUID.fromString("00000002-0000-1000-8000-008025000000");
            case RxCredits:
                return UUID.fromString("00000003-0000-1000-8000-008025000000");
            case TxCredits:
                return UUID.fromString("00000004-0000-1000-8000-008025000000");
            case TxAndTxCredits:
                return UUID.fromString("00002902-0000-1000-8000-00805f9b34fb");
            default:
                throw new IllegalArgumentException("Unknown Telit UUID");
        }
    }

    private TelitUuid toTelitUuid(UUID uuid) {
        if (uuid.equals(toUuid(TelitUuid.Service))) {
            return TelitUuid.Service;
        } else if (uuid.equals(toUuid(TelitUuid.Rx))) {
            return TelitUuid.Rx;
        } else if (uuid.equals(toUuid(TelitUuid.Tx))) {
            return TelitUuid.Tx;
        } else if (uuid.equals(toUuid(TelitUuid.RxCredits))) {
            return TelitUuid.RxCredits;
        } else if (uuid.equals(toUuid(TelitUuid.TxCredits))) {
            return TelitUuid.TxCredits;
        } else if (uuid.equals(toUuid(TelitUuid.TxAndTxCredits))) {
            return TelitUuid.TxAndTxCredits;
        } else {
            throw new IllegalArgumentException("Unknown UUID");
        }
    }

    private void DebugLogBleValues(BluetoothGattCharacteristic characteristic) throws UnsupportedEncodingException {
        if (characteristic == null) {
            return;
        }
        byte[] value = characteristic.getValue();
        if (value == null || value.length == 0) {
            DebugLog("DebugLogBleValues(): Empty value received");
            return;
        }
        TelitUuid telitUuid = toTelitUuid(characteristic.getUuid());
        StringBuilder sb = new StringBuilder();
        sb.append("New value for ");
        sb.append(telitUuid);
        sb.append('\n');
        sb.append("Value as UTF-8 string: ");
        sb.append(new String(value, "UTF-8"));
        sb.append('\n');
        sb.append("Value as 8-bit unsigned integer: ");
        sb.append(characteristic.getIntValue(BluetoothGattCharacteristic.FORMAT_UINT8, 0));
        sb.append('\n');
        sb.append("Value as 16-bit unsigned integer: ");
        sb.append(characteristic.getIntValue(BluetoothGattCharacteristic.FORMAT_UINT16, 0));
        sb.append('\n');
        sb.append("Value as 32-bit unsigned integer: ");
        sb.append(characteristic.getIntValue(BluetoothGattCharacteristic.FORMAT_UINT32, 0));
        sb.append('\n');
        sb.append("Value as 8-bit signed integer: ");
        sb.append(characteristic.getIntValue(BluetoothGattCharacteristic.FORMAT_SINT8, 0));
        sb.append('\n');
        sb.append("Value as 16-bit signed integer: ");
        sb.append(characteristic.getIntValue(BluetoothGattCharacteristic.FORMAT_SINT16, 0));
        sb.append('\n');
        sb.append("Value as 32-bit signed integer: ");
        sb.append(characteristic.getIntValue(BluetoothGattCharacteristic.FORMAT_SINT32, 0));
        sb.append('\n');
        sb.append("Value as 16-bit signed float: ");
        sb.append(characteristic.getFloatValue(BluetoothGattCharacteristic.FORMAT_SFLOAT, 0));
        sb.append('\n');
        sb.append("Value as 32-bit float: ");
        sb.append(characteristic.getFloatValue(BluetoothGattCharacteristic.FORMAT_FLOAT, 0));
        sb.append('\n');
        DebugLog(sb);
    }

    private void DebugLog(@Nullable CharSequence sMessage) {
        if (TextUtils.isEmpty(sMessage)) {
            return;
        }
        Log.v(TAG, sMessage.toString());
    }

    private void handleError(Throwable t) {
        if (t != null) {
            t.printStackTrace();
        }
    }

    private void getTelitCharacteristics() {
        BluetoothGattService service = btGatt.getService(toUuid(TelitUuid.Service));
        if (service == null) {
            throw new NullPointerException("Cannot obtain Telit Terminal IO service");
        }
        charRx = service.getCharacteristic(toUuid(TelitUuid.Rx));
        charTx = service.getCharacteristic(toUuid(TelitUuid.Tx));
        charRxCredits = service.getCharacteristic(toUuid(TelitUuid.RxCredits));
        charTxCredits = service.getCharacteristic(toUuid(TelitUuid.TxCredits));
        if (charRx == null) {
            throw new IllegalStateException("Cannot obtain RX characteristic");
        }
        if (charTx == null) {
            throw new IllegalStateException("Cannot obtain TX characteristic");
        }
        if (charRxCredits == null) {
            throw new IllegalStateException("Cannot obtain RX credits characteristic");
        }
        if (charTxCredits == null) {
            throw new IllegalStateException("Cannot obtain TX credits characteristic");
        }
    }

    @Override
    public void onPhyUpdate(BluetoothGatt btGatt, int bTxPhy, int nRxPhy, int nStatus) {
    }

    @Override
    public void onPhyRead(BluetoothGatt btGatt, int nTxPhy, int nRxPhy, int nStatus) {
    }

    @Override
    public void onConnectionStateChange(BluetoothGatt btGatt, int nStatus, int nNewState) {
        try {
            BluetoothDevice btDevice = btGatt.getDevice();
            String sMacAddress = btDevice.getAddress();
            switch (nNewState) {
                case BluetoothProfile.STATE_DISCONNECTED:
                    DebugLog("Disconnected from device " + sMacAddress);
                    if (!bConnectOperationFinished) {
                        /*
                         * NOTA IMPORTANTE: Este bloque if es un parche guarro para un Samsung S7.
                         * Estos télefonos reciben un STATE_DISCONNECTED justo después del
                         * btDevice.connectGatt(), y la solución es reconectar.
                         *
                         * En teléfonos decentes le llega el STATE_CONNECTED directamente.
                         */
                        DebugLog("BLE disconnected state received, but connect operation didn't finish yet. Retrying connection");
                        boolean bResult = btGatt.connect();
                        if (bResult) {
                            DebugLog("Connection retry operation started successfully");
                        } else {
                            DebugLog("Connection retry operation failed. Clearing flags");
                            this.btGatt = null;
                            this.bConnected = false;
                            this.bConnectOperationFinished = true;
                        }
                        return;
                    }
                    DebugLog("Closing BluetoothGatt object and nullifying references");
                    btGatt.close();
                    /*
                     * Aquí los objetos this.btGatt y btGatt deberían ser la misma instancia siempre.
                     * Anulamos el objeto de la instancia.
                     */
                    this.btGatt = null;
                    this.bConnected = false;
                    this.bConnectOperationFinished = true;
                    break;
                case BluetoothProfile.STATE_CONNECTED:
                    DebugLog("Connected to device " + sMacAddress);
                    List<BluetoothGattService> lstServices = btGatt.getServices();
                    if (lstServices == null || lstServices.size() <= 0) {
                        boolean bResult = btGatt.discoverServices();
                        if (bResult) {
                            DebugLog("Remote device service discovery has been started");
                        } else {
                            DebugLog("Remote device service discovery failed");
                        }
                    } else {
                        DebugLog("Services already found previously. Service count: " + lstServices.size());
                        getTelitCharacteristics();
                        // Primer paso
                        enableTelitTxCredits();
                    }
                    break;
                case BluetoothProfile.STATE_CONNECTING:
                    DebugLog("Connecting to device " + sMacAddress);
                    break;
                case BluetoothProfile.STATE_DISCONNECTING:
                    DebugLog("Disconnecting from device " + sMacAddress);
                    break;
            }
        } catch (Exception ex) {
            /*
             * Sabed que esto nunca ha pasado y que está aquí para evitar posibles cuelgues del
             * framework. Si llega a ocurrir probablemente las banderas de conexión estarán mal.
             */
            handleError(ex);
        }
    }

    @Override
    public void onServicesDiscovered(BluetoothGatt btGatt, int nStatus) {
        try {
            if (nStatus == BluetoothGatt.GATT_SUCCESS) {
                DebugLog("Services discovered successfully");
            } else if (nStatus == BluetoothGatt.GATT_FAILURE) {
                DebugLog("Service discovery failed");
            }
            getTelitCharacteristics();
        } catch (Exception ex) {
            handleError(ex);
            return;
        }
        // Primer paso
        enableTelitTxCredits();
    }

    @Override
    public void onCharacteristicRead(BluetoothGatt btGatt, BluetoothGattCharacteristic characteristic, int nStatus) {
        try {
            TelitUuid telitUuid = toTelitUuid(characteristic.getUuid());
            if (nStatus == BluetoothGatt.GATT_SUCCESS) {
                DebugLog("onCharacteristicRead() OK: " + telitUuid + " status: " + nStatus);
            } else {
                DebugLog("onCharacteristicRead() error: " + telitUuid + " status: " + nStatus);
            }
            byte[] value = characteristic.getValue();
            DebugLog("Value read as UTF-8 string: " + new String(value, "UTF-8"));
        } catch (Exception ex) {
            handleError(ex);
        }
    }

    @Override
    public void onCharacteristicWrite(BluetoothGatt btGatt, BluetoothGattCharacteristic characteristic, int nStatus) {
        try {
            TelitUuid telitUuid = toTelitUuid(characteristic.getUuid());
            if (nStatus == BluetoothGatt.GATT_SUCCESS) {
                DebugLog("onCharacteristicWrite() OK: " + telitUuid + " status: " + nStatus);
                if (telitUuid == TelitUuid.RxCredits) {
                    DebugLog("Credits characteristic written");
                }
            } else {
                throw new IOException("onCharacteristicWrite() error: " + telitUuid + " status: " + nStatus);
            }
        } catch (Exception ex) {
            handleError(ex);
        }
    }

    @SuppressWarnings("StatementWithEmptyBody")
    @Override
    public void onCharacteristicChanged(BluetoothGatt btGatt, BluetoothGattCharacteristic characteristic) {
        try {
            TelitUuid telitUuid = toTelitUuid(characteristic.getUuid());
            DebugLog("onCharacteristicChanged(): " + telitUuid);
            DebugLogBleValues(characteristic);
            if (telitUuid == TelitUuid.TxCredits) {
                nReceivedCredits = characteristic.getIntValue(BluetoothGattCharacteristic.FORMAT_UINT8, 0);
                bConnected = true;
                bConnectOperationFinished = true;
            } else if (telitUuid == TelitUuid.Tx) {
                // Respuestas del servidor. Las recogemos con getCachedResponse();
            }
        } catch (Exception ex) {
            handleError(ex);
        }
    }

    @Override
    public void onDescriptorRead(BluetoothGatt btGatt, BluetoothGattDescriptor descriptor, int nStatus) {
        TelitUuid telitUuid = toTelitUuid(descriptor.getUuid());
        if (nStatus == BluetoothGatt.GATT_SUCCESS) {
            DebugLog("onDescriptorRead() OK: " + telitUuid + " status: " + nStatus);
        } else {
            DebugLog("onDescriptorRead() error: " + telitUuid + " status: " + nStatus);
        }
    }

    @SuppressWarnings("StatementWithEmptyBody")
    @Override
    public void onDescriptorWrite(BluetoothGatt btGatt, BluetoothGattDescriptor descriptor, int nStatus) {
        try {
            TelitUuid telitUuid = toTelitUuid(descriptor.getUuid());
            if (nStatus == BluetoothGatt.GATT_SUCCESS) {
                DebugLog("onDescriptorWrite() OK: " + telitUuid + " status: " + nStatus);
                if (Arrays.equals(descriptor.getValue(), BluetoothGattDescriptor.ENABLE_INDICATION_VALUE)) {
                    enableTelitTx();
                } else if (Arrays.equals(descriptor.getValue(), BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE)) {
                    sendCredits();
                } else {
                    //A saber...
                }
            } else {
                throw new IOException("onDescriptorWrite() error: " + telitUuid + " status: " + nStatus);
            }
        } catch (Exception ex) {
            handleError(ex);
        }
    }

    @Override
    public void onReliableWriteCompleted(BluetoothGatt btGatt, int nStatus) {
        if (nStatus == BluetoothGatt.GATT_SUCCESS) {
            DebugLog("onReliableWriteCompleted() OK. Status: " + nStatus);
        } else {
            DebugLog("onReliableWriteCompleted() error. Status: " + nStatus);
        }
    }

    @Override
    public void onReadRemoteRssi(BluetoothGatt btGatt, int nRssi, int nStatus) {
    }

    @Override
    public void onMtuChanged(BluetoothGatt btGatt, int nMtu, int nStatus) {
    }

    /**
     * Primer paso, registrar este descriptor con indicaciones BLE para recibir los créditos.
     */
    private void enableTelitTxCredits() {
        boolean bResult = btGatt.setCharacteristicNotification(charTxCredits, true);
        DebugLog("Enabling TX credits characteristic notification: " + bResult);
        BluetoothGattDescriptor charTxCreditsDescriptor = charTxCredits.getDescriptor(toUuid(TelitUuid.TxAndTxCredits));
        charTxCreditsDescriptor.setValue(BluetoothGattDescriptor.ENABLE_INDICATION_VALUE);
        bResult = btGatt.writeDescriptor(charTxCreditsDescriptor);
        DebugLog("Enabling remotely TX credits characteristic indication: " + bResult);
    }

    /**
     * Segundo paso, después de recibir las notificaciones de créditos, nos suscribimos a las
     * notificaciones de respuesta, para recibir datos.
     */
    private void enableTelitTx() {
        boolean bResult = btGatt.setCharacteristicNotification(charTx, true);
        DebugLog("Enabling TX characteristic notification: " + bResult);
        BluetoothGattDescriptor charTxDescriptor = charTx.getDescriptor(toUuid(TelitUuid.TxAndTxCredits));
        charTxDescriptor.setValue(BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE);
        bResult = btGatt.writeDescriptor(charTxDescriptor);
        DebugLog("Enabling remotely TX characteristic notification: " + bResult);
    }

    /**
     * Tercer paso, enviar los créditos. El 255 aquí es el valor máximo. Si se realizasen más de 255
     * operaciones, esto ya dejaría de funcionar. Eso por ahora no está contemplado.
     */
    private void sendCredits() throws IOException {
        DebugLog("Sending 25 credits");
        doWriteUint8Internal(charRxCredits, 25);
    }

    @SuppressWarnings("unused")
    public int getReceivedCredits() {
        return nReceivedCredits;
    }

    @SuppressWarnings("SameParameterValue")
    private void doWriteUint8Internal(BluetoothGattCharacteristic characteristic, int nValue) throws IOException {
        if (btGatt == null) {
            throw new IOException("BLE device not connected");
        }
        // Borrar la respuesta cacheada de alguna orden previa
        charTx.setValue((byte[]) null);
        TelitUuid telitUuid = toTelitUuid(characteristic.getUuid());
        DebugLog("writeUint8 for characteristic " + telitUuid);
        boolean bResult = characteristic.setValue(nValue, BluetoothGattCharacteristic.FORMAT_UINT8, 0);
        DebugLog("setValue(): " + bResult);
        bResult = btGatt.writeCharacteristic(characteristic);
        DebugLog("writeCharacteristic(): " + bResult);
    }

    private void doWriteStringInternal(BluetoothGattCharacteristic characteristic, String sValue) throws IOException {
        if (btGatt == null) {
            throw new IOException("BLE device not connected");
        }
        // Borrar la respuesta cacheada de alguna orden previa
        charTx.setValue((byte[]) null);
        TelitUuid telitUuid = toTelitUuid(characteristic.getUuid());
        DebugLog("writeString for characteristic " + telitUuid);
        byte[] buffer = (sValue + "\r\n").getBytes("UTF-8");
        buffer = Arrays.copyOf(buffer, 20);
        boolean bResult = characteristic.setValue(buffer);
        DebugLog("setValue(): " + bResult);
        bResult = btGatt.writeCharacteristic(characteristic);
        DebugLog("writeCharacteristic(): " + bResult);
    }

    public void writeUint8(int nValue) throws IOException {
        doWriteUint8Internal(charRx, nValue);
    }

    public void writeString(String sValue) throws IOException {
        doWriteStringInternal(charRx, sValue);
    }

    public String getCachedResponse() {
        return charTx.getStringValue(0);
    }

    public void connect(@NonNull String sMacAddress, long nTimeout, boolean bAsync, boolean bAutoConnect) throws InterruptedException, IOException {
        this.sMacAddress = sMacAddress;
        this.nConnectTimeout = nTimeout;
        this.bAutoConnect = bAutoConnect;
        if (bAsync) {
            connectAsync();
        } else {
            connectSync();
        }
    }

    private void connectSync() throws InterruptedException, IOException {
        connectAsync();
        blockUntilConnected(true);
    }

    private void connectAsync() {
        DebugLog("Initiating new BLE connection to device " + sMacAddress);
        if (Looper.myLooper() == Looper.getMainLooper()) {
            throw new IllegalStateException("BleSerialPortTelitGattCallback.connectSync() cannot be called from the UI thread");
        }
        if (TextUtils.isEmpty(sMacAddress)) {
            throw new IllegalArgumentException("Empty MAC address");
        }
        BluetoothDevice btDevice = btAdapter.getRemoteDevice(sMacAddress);
        if (btDevice == null) {
            throw new IllegalStateException("Cannot obtain bluetooth device with MAC address " + sMacAddress);
        }
        bConnected = false;
        bConnectOperationFinished = false;
        btGatt = btDevice.connectGatt(context, bAutoConnect, this);
        if (btGatt == null) {
            throw new IllegalStateException("Cannot obtain bluetooth GATT device with MAC address " + sMacAddress);
        }
    }

    public void blockUntilConnected(boolean bInitialConnection) throws InterruptedException, IOException {
        if (bInitialConnection || !bConnectOperationFinished) {
            for (long i = 0; !bConnectOperationFinished && i < nConnectTimeout; i = i + 50) {
                TimeUnit.MILLISECONDS.sleep(50);
            }
            // Ni siquiera el último callback que marca bConnectOperationFinished = true es fiable.
            TimeUnit.MILLISECONDS.sleep(500);
        }
        if (!bConnected) {
            throw new IOException("BLE connection to device " + sMacAddress + " failed");
        }
    }

    public void disconnect() {
        bConnectOperationFinished = true;
        if (btGatt != null) {
            DebugLog("Disconnecting from device " + btGatt.getDevice().getAddress());
            btGatt.disconnect();
        }
    }
}
```

**Comment by The-Mekon** [comment]
Thanks so much for getting back so quickly, and with so much useful code.  If I can work through that, I'm sure I'll get there.

I'm actually coding my app in JS, using the experimental Web Bluetooth API .  I guess I just like to live dangerously! :)

**Comment by kai-morich** [comment]
I will have a look, if I can also add the variant with telit communication to this repo

**Comment by The-Mekon** [comment]
Thanks Kai - that could be very useful.  There doesn't seem to be much out there that will actually talk to Telit devices.

**Comment by kai-morich** [comment]
implemented with 27f1b1fda7162fc2218d98e278373c9c5539df98
